### PR TITLE
etcd: 3.4.16 -> 3.4.18, 3.3.25 -> 3.3.27, init 3.5.1

### PR DESCRIPTION
**Pull request by bryanasdev000**
###### Motivation for this change

Update etcd versions

WIP on 3.5.1

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by bryanasdev000** [comment]
At the moment I can't build 3.5.1, if I set vendorSha256 it fails trying to go get with git:

```
DEPRECATED!!! Use build.sh script instead.

% 'env' 'GO111MODULE=off' 'go' 'get' 'github.com/myitcv/gobin'
stderr: go: missing Git command. See https://golang.org/s/gogetcmd
stderr: package github.com/myitcv/gobin: exec: "git": executable file not found in $PATH
FAIL: (code:1):
  % 'env' 'GO111MODULE=off' 'go' 'get' 'github.com/myitcv/gobin'
```

If I keep the same style as `3.4.nix` it complains about `go mod vendor`:

```
DEPRECATED!!! Use build.sh script instead.

go: inconsistent vendoring in /build/source:
        github.com/bgentry/speakeasy@v0.1.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        github.com/dustin/go-humanize@v1.0.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        github.com/spf13/cobra@v1.1.3: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        go.etcd.io/bbolt@v1.3.6: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        go.etcd.io/etcd/api/v3@v3.5.1: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        go.etcd.io/etcd/client/pkg/v3@v3.5.1: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        go.etcd.io/etcd/client/v2@v2.305.1: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        go.etcd.io/etcd/client/v3@v3.5.1: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        go.etcd.io/etcd/etcdctl/v3@v3.5.1: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        go.etcd.io/etcd/etcdutl/v3@v3.5.1: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        go.etcd.io/etcd/pkg/v3@v3.5.1: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        go.etcd.io/etcd/raft/v3@v3.5.1: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        go.etcd.io/etcd/server/v3@v3.5.1: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        go.etcd.io/etcd/tests/v3@v3.5.1: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        go.uber.org/zap@v1.17.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        golang.org/x/time@v0.0.0-20210220033141-f8bda1e9f3ba: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        google.golang.org/grpc@v1.38.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        gopkg.in/cheggaaa/pb.v1@v1.0.28: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        go.etcd.io/etcd/api/v3: is replaced in go.mod, but not marked as replaced in vendor/modules.txt
        go.etcd.io/etcd/client/pkg/v3: is replaced in go.mod, but not marked as replaced in vendor/modules.txt
        go.etcd.io/etcd/client/v2: is replaced in go.mod, but not marked as replaced in vendor/modules.txt
        go.etcd.io/etcd/client/v3: is replaced in go.mod, but not marked as replaced in vendor/modules.txt
        go.etcd.io/etcd/etcdctl/v3: is replaced in go.mod, but not marked as replaced in vendor/modules.txt
        go.etcd.io/etcd/etcdutl/v3: is replaced in go.mod, but not marked as replaced in vendor/modules.txt
        go.etcd.io/etcd/pkg/v3: is replaced in go.mod, but not marked as replaced in vendor/modules.txt
        go.etcd.io/etcd/raft/v3: is replaced in go.mod, but not marked as replaced in vendor/modules.txt
        go.etcd.io/etcd/server/v3: is replaced in go.mod, but not marked as replaced in vendor/modules.txt
        go.etcd.io/etcd/tests/v3: is replaced in go.mod, but not marked as replaced in vendor/modules.txt

        To ignore the vendor directory, use -mod=readonly or -mod=mod.
        To sync the vendor directory, run:
                go mod vendor
must be run from 'go.etcd.io/etcd/v3' module directory
```

Also, I am not sure how to run NixOS tests.

**Comment by thiagokokada** [comment]
> Also, I am not sure how to run NixOS tests.

https://nixos.org/manual/nixos/unstable/index.html#sec-running-nixos-tests

**Comment by bryanasdev000** [comment]
> > Also, I am not sure how to run NixOS tests.
> 
> https://nixos.org/manual/nixos/unstable/index.html#sec-running-nixos-tests

Thanks! Both tests is ok.

But I think we need to modify the current module to use a different version that default (v3.3).

**Comment by bryanasdev000** [comment]
If changing default version to 3.4.X or 3.5.X we will need to upgrade tests.

File: pkgs/servers/etcd/3.4.nix [hunk 1]
```diff
@@ -1,8 +1,8 @@
-{ lib, buildGoModule, fetchFromGitHub }:
+{ lib, buildGoModule, fetchFromGitHub, nixosTests }:
```
Comment (zowoq) [inline pkgs/servers/etcd/3.4.nix:1]:
```suggestion
{ lib, buildGoModule, fetchFromGitHub }:
```

File: pkgs/servers/etcd/3.4.nix [hunk 2]
```diff
@@ -26,6 +26,8 @@ buildGoModule rec {
     install -Dm755 bin/* bin/functional/cmd/* -t $out/bin
   '';
 
+  passthru.tests = { inherit (nixosTests) etcd etcd-cluster; };
```
Comment (zowoq) [inline pkgs/servers/etcd/3.4.nix:29]: Pointless to add this as it is a no op. Only `etcd` is tested, not the versioned packages.

The test will also fail for >= 3.4.

```suggestion
```

File: pkgs/servers/etcd/3.5.nix [hunk 3]
```diff
@@ -0,0 +1,42 @@
+{ lib, buildGoModule, fetchFromGitHub, nixosTests }:
```
Comment (zowoq) [inline pkgs/servers/etcd/3.5.nix:1]:
```suggestion
{ lib, buildGoModule, fetchFromGitHub }:
```

File: pkgs/servers/etcd/3.5.nix [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+buildGoModule rec {
+  pname = "etcd";
+  version = "3.5.1";
+
+  #deleteVendor = true;
```
Comment (zowoq) [inline pkgs/servers/etcd/3.5.nix:7]:
```suggestion
```

File: pkgs/servers/etcd/3.5.nix [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+  vendorSha256 = "sha256-ad89jDLLeAJizQMmO0lPH8N9GdaBrn9NyXGCxVT3fHE=";
+  #runVend = true;
```
Comment (zowoq) [inline pkgs/servers/etcd/3.5.nix:9]:
```suggestion
```

File: pkgs/servers/etcd/3.5.nix [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+  #preBuild = ''
+  #  env go generate ./build/source
+  #'';
```
Comment (zowoq) [inline pkgs/servers/etcd/3.5.nix:13]:
```suggestion
```

File: pkgs/servers/etcd/3.5.nix [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+    patchShebangs .
+    ./build
+    ./functional/build
+  '';
+
+  installPhase = ''
+    install -Dm755 bin/* bin/functional/cmd/* -t $out/bin
+  '';
+
+  passthru.tests = { inherit (nixosTests) etcd etcd-cluster; };
```
Comment (zowoq) [inline pkgs/servers/etcd/3.5.nix:33]: As above.
```suggestion
```

**Review by zowoq — CHANGES_REQUESTED** [review]
Commits should amended to follow https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#submitting-changes.

`etcd: 3.3.25 -> 3.3.27`
`etcd_3_4: 3.3.25 -> 3.3.27`
`etcd_3_5: init at 3.5.1`
`nixos/etcd: add option to set package`

File: pkgs/servers/etcd/3.4.nix (hunk 2 above)
Comment (bryanasdev000) [reply pkgs/servers/etcd/3.4.nix:29]: What about other versions?

I can give a look at tests and update it too.

Comment (bryanasdev000) [reply pkgs/servers/etcd/3.4.nix:29]: Also, is it worth leaving 3.5.1 as the default version?

**Comment by bryanasdev000** [comment]
> Commits should amended to follow https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#submitting-changes.
> 
> `etcd: 3.3.25 -> 3.3.27` `etcd_3_4: 3.3.25 -> 3.3.27` `etcd_3_5: init at 3.5.1` `nixos/etcd: add option to set package`

@zowoq  What about other versions? is it worth leaving 3.5.1 as the default version?

Also I can give a look at tests and update it too.

**Comment by zowoq** [comment]
> is it worth leaving 3.5.1 as the default version?

Besides fixing the `etcd` tests would need to make sure it works with everything else that depends on it, e.g. the kubernetes modules.

**Comment by bryanasdev000** [comment]
> > is it worth leaving 3.5.1 as the default version?
> 
> Besides fixing the `etcd` tests would need to make sure it works with everything else that depends on it, e.g. the kubernetes modules.



> > is it worth leaving 3.5.1 as the default version?
> 
> Besides fixing the `etcd` tests would need to make sure it works with everything else that depends on it, e.g. the kubernetes modules.

Hmm, this will be a little rabbit hole.

Can I set package on tests and make it test the latest release?

Any tip with 3.5.1 build?

File: nixos/modules/services/misc/etcd.nix [hunk 8]
```diff
@@ -184,7 +190,7 @@ in {
       };
     };
 
-    environment.systemPackages = [ pkgs.etcd ];
+    environment.systemPackages = [ cfg.package ];
```
Comment (zowoq) [inline nixos/modules/services/misc/etcd.nix:193]: This needs to be updated in the `serviceConfig` as well.

Comment (bryanasdev000) [reply nixos/modules/services/misc/etcd.nix:193]: Done.

**Comment by zowoq** [comment]
@bryanasdev000 Might be easier to split up this PR into three.

`etcd` and `etcd_3_4` version bumps could be merged immediately.

`etcd_3_5` looks like it's needs some changes to the build process.

`nixos/etcd` + tests + changing default version looks much more complicated with breaking changes that will need to be handled.

**Comment by bryanasdev000** [comment]
> @bryanasdev000 Might be easier to split up this PR into three.
> 
> `etcd` and `etcd_3_4` version bumps could be merged immediately.
> 
> `etcd_3_5` looks like it's needs some changes to the build process.
> 
> `nixos/etcd` + tests + changing default version looks much more complicated with breaking changes that will need to be handled.

No problem, I will do it as soon as I have time. I will close this one for now and leave a comment with the new PRs.

Thanks.

**Comment by zowoq** [comment]
I cherry picked the version bumps to https://github.com/NixOS/nixpkgs/pull/146040

**Comment by bryanasdev000** [comment]
@zowoq here it is:

- https://github.com/NixOS/nixpkgs/pull/146331 (nixos module and tests)
- https://github.com/NixOS/nixpkgs/pull/146327 (etcd_3_5)

**Pull request closed without merging**